Fix for issue #1277

nicolasbeauvais opened the pull request:
This PR make two changes:

### Minor: 
The `performConversion` method only perform manipulations has the conversion is made in the `performConversions` method [on line 94](https://github.com/spatie/laravel-medialibrary/compare/master...nicolasbeauvais:master#diff-0d11c445b5c28ccade90cf390e511012L94). 
So it makes more sense to rename it `performManipulations`instead. It will also reduce confusion between `performConversions` and `performConversion`.

### Fix for issue #1277:

#### Problem:
Creating a custom generator that returns a file that is not an image (in that case a .m3u8 file) fail because manipulations are being automatically run on that file which is not compatible.

#### Proposed fix:
The end user needs to disable the default manipulations for the specific conversion like:
```php
$this->addMediaConversion('transcode')
            ->removeManipulation('format')
            ->removeManipulation('optimize')
            ->performOnCollections('file');

// We could also add a base method for this:

$this->addMediaConversion('transcode')
            ->removeAllManipulations()
            ->performOnCollections('file');
```

And in the core, we check if the running conversion has manipulation to be triggered, if not we skip this step and the converted file stay untouched.


I haven't dive in the core code for a long time so I might be missing something here, anyway do not hesitate to indicate any change that could be made to improve this fix.

pedrofurtado commented:
@nicolasbeauvais The CI failed because of some rules of coding standard 👍

pedrofurtado commented:
Ping for notice: @freekmurze

freekmurze commented on src/FileManipulator.php:
```diff
@@ -93,13 +93,17 @@ public function performConversions(ConversionCollection $conversions, Media $med
-                $conversionResult = $this->performConversion($media, $conversion, $copiedOriginalFile);
+                if (! $conversion->getManipulations()->isEmpty()) {
+                    $manipulationResult = $this->performManipulations($media, $conversion, $copiedOriginalFile);
+                } else {
```
Could your refactor this so we don't need an `else` statement? A dedicated method with an early return seems the way to go.

freekmurze commented:
@nicolasbeauvais Thanks for this! 👍 Could you add a test to make sure that these changes work?

nicolasbeauvais commented:
WIP:
- [x] Refactored the fix using an early return in the `performManipulations` method.
- [x] Created a `withoutManipulations` helper function for conversions.
- [x] Added some missing return types in `Conversion.php`.
- [ ] Add tests for the fix.
- [ ] Add tests for the `withoutManipulations` method.
- [ ] Update the documentation to reflect this changes

nicolasbeauvais commented:
I should be done with this, of course, feel free to suggest any improvements.

I will update the documentation once this gets merged!

freekmurze commented:
Thank you so much for your work on this 👍

chelout commented:
@nicolasbeauvais @pedrofurtado could you please add an example who to work with this feature?

sp4r74cus commented:
hi @pedrofurtado, can you confirm your hls conversion works with this update? i tried to make a similar conversion (mp4 -> hls) using your code as a start ([#1277](https://github.com/spatie/laravel-medialibrary/issues/1277#issuecomment-431634011)) 

i'm facing 2 issues:

1. the conversion file is saved as video-{conversion-name}.mp4 and not .m3u8 (the file name i return)
2. all the .m3u8 & .ts generated in the tmp directory are not moved to the conversions folder 

Thanks

jayenne commented:
did the two issues above regarding conversio nfile name and moving t ocorrect dir get resolved?
- would you be able to share a howto for converting video/audio media to .m3u8 & .ts please?
- or point us to a guide/resource please?

joserick commented:
Hi all.

`the conversion file is saved as video-{conversion-name}.mp4 and not .m3u8 (the file name i return)`

I had the first problem @jayenne , I solved it in the following way:

```php
$this->addMediaCollection('audio')->singleFile()
    ->registerMediaConversions(function (Media $media) {
        $this->addMediaConversion('ogg')
            //->withoutManipulations() #Removed for redundancy in this case.
            ->setManipulations(function (Manipulations &$manipulations){
                $manipulations = new class extends Manipulations{
                    const FORMAT_OGG = 'ogg';

                    public function isEmpty(): bool
                    {
                        return true;
                    }
                };

                $manipulations->format('ogg');
            })->nonQueued();
    });
```
```php
MyModel::first()->addMedia('path/to/file.wav')->toMediaCollection('audio');
```
```php
MyModel::first()->getFirstMediaUrl('audio', 'ogg');
// http://localhost/storage/#/conversions/file-ogg.ogg <- Already change the extension
```

I hope it will help someone in the future :)
